### ISSUE TITLE
Add a render mode to allow depth testing and screen reading transparent objects.

### ISSUE THREAD

**Issue by Mopzilla** [godotengine/godot-proposals#10847]
### Describe the project you are working on

I am working on a project that heavily uses transparency

### Describe the problem or limitation you are having in your project

Transparency is very difficult to work with - especially when they overlap. I have issues with transparent objects blending improperly with each other, and with the sky texture.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a render_mode flag that allows a shader to depth test transparent shaders and include transparent shaders in screen reads by reading after the transparent layer.

This way on my screen space shader I can accurately get the depth/screen of every object in the scene.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I am unsure how it would work in the engine, but my assumption is performance would only be lost when actually using the render mode.

For example add render modes: `depth_test_alpha` and `screen_read_alpha`
Using depth_test_alpha you have transparent objects included in your depth texture
Using screen_read_alpha you have transparent objects included in your screen texture

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I can't actually think of any solution whatsoever, which is why I am writing this. If there were a work-around that was somewhat convenient I would use it (if one exists let me know!)

### Is there a reason why this should be core and not an add-on in the asset library?

I can't think of any downsides except performance which only happens if you actively use the render mode. And there are so many flaws with transparency handling in-engine, so it would be nice to have it built in.

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot-proposals/issues/7916.

I don't know if this is technically feasible given how Godot's renderer works, especially since it currently only supports forward rendering and not [deferred rendering](https://github.com/godotengine/godot-proposals/issues/8295).

**Comment by Mopzilla** [comment]
I see, before I posted this proposal I did look at some other posts and someone said they just decided to do it themselves in the engine so I figured it was something easy to implement and was missing for some other reasons. At least the screen reading that is.

**Comment by Jesusemora** [comment]
We many options for transparency. You have to be more specific with what you want and what your problem is, because it could be a genuine limitation, or you just haven't done your reseach.

**Comment by Mopzilla** [comment]
I want a screen effect that occurs when the camera is contained within a water volume which distorts the screen, as well as layering other effects.

The issue is I cannot read the screen texture to distort it, as transparent shaders are not included in the screen texture.

I also cannot depth test the water volume to determine how thick the fog effect should be, as you cannot depth test transparent shaders.

I think there is a solution with using the compositor, but that is far above my current knowledge (though I am trying to understand it).

If there was something simple like a render mode that exposed this functionality then it would make creating these effects extremely easy for less experienced developers.

**Comment by PureAsbestos** [comment]
You want a post-processing effect for this. See [this page in the docs](https://docs.godotengine.org/en/stable/tutorials/shaders/advanced_postprocessing.html) for more info. You don't actually need to get the depth of transparent objects to do this (unless you have transparent things underwater maybe, but that's a separate problem).

**Comment by Mopzilla** [comment]
That docs page is what I used to build the shader. The issue is (again this could be the wrong way to go about what I want) inside my quad shader that covers the camera I want to read the depth so I can determine how transparent a given fragment should be.

Objects up close are 0 ALPHA, and objects further than a certain range are 1 ALPHA.

The problem arises because I cannot get the depth of transparent materials, so if the only thing between my camera and the sky (or an object far away) is a transparent waterfall, it gets covered because the quad shader displays 1 ALPHA.

If the surface of the waterfall shader was included in the depth test then my quad screen shader would render at 0.2 ALPHA on that fragment, allowing the waterfall to be visible.

I am very new to shader programming and this could be the wrong way to go about what I want to achieve. (screen effect which adds fake fog, the further away a fragment is from the camera the higher the alpha value for that fragment)

**Comment by thompsop1sou** [comment]
I've also worked with screen-reading effects that both needed to use the depth texture and needed to be able to see transparent objects. Godot doesn't have great options for this right now (as you've discovered), but they are already currently in development (see the [Rendering Compositor issue](https://github.com/godotengine/godot-proposals/issues/7916), which Calinou mentioned above).

If you need a way to work around this right now (before the Rendering Compositor is finished), here are a few options:

1. If you only need to access the color texture (`hint_screen_texture`) and not the depth texture or the normal-roughness texture, then you can just use a 2D screen-reading effect. You'll put this on a full-screen `TextureRect` or `Sprite2D`. Because this is a 2D effect, *all* of the 3D rendering will be already completed in the texture that you access using `hint_screen_texture`, including rendering of transparent objects. In addition, you can stack screen-reading effects using the `BackBufferCopy` node (see the docs [here](https://docs.godotengine.org/en/stable/tutorials/shaders/screen-reading_shaders.html#behind-the-scenes)).

2. If you're willing to do a little low-level code (compute shaders, GLSL, and using the `RenderingDevice`), you can use the already completed `Compositor` and `CompositorEffect` resources. (These were introduced in Godot 4.3.) This allows you to add a post-processing effect at several points in the 3D rendering pipeline, including *after* transparent objects have been rendered. You can access any of the screen textures that you might use in a normal shader (`hint_screen_texture`, `hint_depth_texture`, or `hint_normal_roughness_texture`). However, you'll have to write the low-level code to explicitly pass these to a compute shader. The docs go over a complete example of how this works [here](https://docs.godotengine.org/en/stable/tutorials/rendering/compositor.html).

> **Note:** Even though a compositor effect does have access to all of the screen textures *after* the transparent pass, transparent will only automatically write to the color texture. To get them to write to the depth texture, you'll need to set them to `depth_draw_always`. AFAIK, there is no way to get them to draw to the normal texture (see [this issue](https://github.com/godotengine/godot/issues/93217)).

3. There's also a way to pass the screen textures about manually via viewports, but it takes quite a bit of setup. However, the setup doesn't really involve low-level code, so this might be a better option if you'd prefer to stay away from that. In general, you'll do something like the following:
   1. You'll need two cameras. One will capture color information (so we'll call it the "color camera") and the other will capture depth information (so we'll call it the "depth camera"). You'll want the two cameras to have all of the same properties *except* that they should be on different visual layers. You'll also need to make it so that their 3D transforms match (you could make the depth camera child of the color camera).
   2. Add two `SubViewport` nodes to your setup. You'll want to hook up one viewport to the color camera and the other viewport to the depth camera. You can do this by writing a little script (can go on the cameras or on the viewports). You'll make use of the method `RenderingServer.viewport_attach_camera()` (see [here](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#class-renderingserver-method-viewport-attach-camera)).
   3. Pass the resulting textures from these viewports as uniforms to your post-processing shader. See [here](https://docs.godotengine.org/en/stable/tutorials/shaders/using_viewport_as_texture.html) for an example of how to do this. After this is done, you will access the color texture and depth texture via these uniforms. (You will no longer need to use `hint_screen_texture` and `hint_depth_texture`. Basically, you've manually passed in these textures instead of using Godot's built-in versions.)
   5. Finally, you'll need to write a custom shader for *all* the objects that you want to be correctly visible in these textures that you just created. In this shader (which, again, needs to be put on all your 3D objects), you'll overwrite the `fragment()` function so that it outputs regular color information for the color camera and then outputs depth information for the depth camera. You can do this by testing which layer the current camera is on using the shader built-in `CAMERA_VISIBLE_LAYERS`. If it's on the color layer, output regular color; if it's on the depth layer, output depth information. (I think this built-in is only available in the `vertex()` function by default, but you can pass it as a `varying` to the `fragment()` function.)

> If you decide to go with this third option, I'd be happy to provide more details. I've done it before, so I know it can work. But it is a little finicky to get everything set up correctly.

**Comment by Mopzilla** [comment]
Wow that is a beautiful write up, and helps immensely. I have been putting off learning compute shaders to get the effect I want and have been working on other parts of my game. I definitely think solution 3 seems the most convenient for me as I am still hesitant on delving into the compositor.

**Comment by thompsop1sou** [comment]
I decided to go ahead and make a simple project showcasing the third method (passing screen textures via viewports). It's copied from another project I had been working on and stripped down to its essentials. (Although, I did include several material properties on the object shader which might not be essential, such as metallic, roughness, and emission.) Hopefully this can be helpful for you!

https://github.com/thompsop1sou/custom-screen-buffers

**Comment by Mopzilla** [comment]
Thank you, that's very helpful

**Comment by PureAsbestos** [comment]
> I am very new to shader programming and this could be the wrong way to go about what I want to achieve. (screen effect which adds fake fog, the further away a fragment is from the camera the higher the alpha value for that fragment)

Yeah so the easiest way to do that especially if you only need it on the waterfall is actually probably to just have a special material on the waterfall that fades as the player gets closer (i.e. using `CAMERA_POSITION_WORLD` and the frag coord)

**Comment by romgerman** [comment]
This is still a huge problem today if you want to make a post-processing effect which uses a depth buffer and you also want your transparent objects to show up on your screen at the same time. All of the workarounds are non-trivial and have downsides - you either write a low level compositor effect or write your own shader/camera system from scratch.

My problem right now is that I've followed the "Advanced post-processing" tutorial (which btw doesn't even tell you that your transparent objects are gone after you make your full-screen quad shader) and my 3D UI (using a Sprite3D) is gone. It's showing up only if I disable my full-screen quad post-processing node (the full-screen quad also breaks the viewport gizmos inside the editor). I haven't found any other workarounds except what **thompsop1sou** wrote.

**Comment by wagnerfs** [comment]
Honestly been stuck on the same situation as @romgerman, I need both screen_texture with transparent objects and depth_texture on compatibility mode, canvas shader can only do screen texture, spatial can only do depth, compositors aren't available on compatibility.

@thompsop1sou 3rd setup is what I tried before coming in here as a last resort, but rendering the depth on a viewport and using it on a canvas shader, absolutely worked but cut my FPS in half.

By now I'm probably giving up of it and hoping for either compositor on compatibility or some sort of backbuffercopy for 3D or anything regarding both depth and screen textures post transparent objects in a single shader type.

**Comment by romgerman** [comment]
> By now I'm probably giving up of it and hoping for either compositor on compatibility or some sort of backbuffercopy for 3D or anything regarding both depth and screen textures post transparent objects in a single shader type

For now I resolved this issue this way (which is still not ideal but works for me. I use Forward+):
1. I use two additional viewports (think of them as the "second viewport"): one for rendering, another one for depth rendering. Yes it renders everything twice but I haven't gotten into perf problems yet. Both use [AttachableSubViewport](https://github.com/thompsop1sou/custom-screen-buffers/blob/main/objects/attachable_sub_viewport.gd) from the thompsop1sou's solution.
2. I decided to use Compositor + this addon https://github.com/DevPrice/godot-slang which allows me to ditch the boilerplate code for the Compositor (the default way when you write glsl and your own gdscript wrapper). One con is that you need to learn the slang shader language a bit, but it's very forgiving if you are used to glsl/hlsl/gdshader.

<details>
<summary>Details</summary>

<img width="307" height="195" alt="Image" src="https://github.com/user-attachments/assets/6d531444-0dba-45c3-8465-406313dc6663" />

^ here "DiffuseCamera" is for rendering your "second viewport". "DepthCamera" is for rendering depth: it contains a simple compute shader which writes contents of the viewport's depth buffer (which you can access only when using the Compositor) to the screen texture of the viewport (I haven't gotten into depth packing or anything that's why it linearizes the depth before writing it to the texture, otherwise it won't fit):

```slang
import godot;

[gd::compositor::ColorTexture]
let out_color : RWTexture2D<float4>;

[gd::compositor::DepthTexture]
let depth_texture : Sampler2D;

[gd::compositor::InternalSize]
let viewport_size : int2;

[gd::compositor::SceneData]
let scene_data : ConstantBuffer<gd::compositor::SceneDataBlock>;

[shader("compute")]
[numthreads(8, 8, 1)]
func display_depth_and_stuff(uint3 threadId: SV_DispatchThreadID) {
	let coord = int2(threadId.xy);
    if (coord.x >= viewport_size.x || coord.y >= viewport_size.y) return;

    let screen_uv = float2(coord + .5f) / viewport_size;
	let depth = linear_depth(depth_texture, screen_uv, scene_data.data.inv_projection_matrix);

	out_color[threadId.xy] = depth;
}

float linear_depth(Sampler2D depth_tex, float2 uv, float4x4 inv_projection) {
    let depth = depth_tex.Sample(uv).r;
    let ndc = float3(uv * 2.0 - 1.0, depth);
    var view = mul(float4(ndc, 1.0), inv_projection);
    view.xyz /= view.w;
    return -view.z;
}
```

After that the main camera can use both viewport textures inside a compute shader where you do all of your stuff.

<img width="375" height="337" alt="Image" src="https://github.com/user-attachments/assets/72795422-b41b-40ed-a0e3-b111bbd61b29" />

```slang
// main viewport color
[gd::compositor::ColorTexture]
RWTexture2D<float4> scene_color;

// main viewport depth
[gd::compositor::DepthTexture]
let depth_texture : Sampler2D;

// depth from the DepthViewport
[gd::Export]
[gd::DefaultWhite]
let depth_viewport_texture : Sampler2D;

// color from the DiffuseViewport
[gd::Export]
let diffuse_viewport_texture : Sampler2D;

// blah blah blah your shader is here
```

</details>

**Comment by wagnerfs** [comment]
> I use two additional viewports (think of them as the "second viewport"):

Yeah, using Viewports is a heavy hit on the frame rate, thompsop1sou's approach works, but it's far from ideal, you're still re-rendering the scene again only to extract data, the example project alone (a still scene) renders at 1600 FPS for me, with the camera trick off, it goes up to 2600, and that's a still scene, I only had one in my tests just for the depth and from a steady 160 it went to 70, when you think people not able to play my game because vulkan fallback to gles3 having the fps tank like that is a no no.

For forward+ I can get away with just Compostiros on a post-transparent effect callback, I don't mind writing glsl if it means I get access to post-transparent post-processing on gles3.

Tbh, right now the most sensible way to have easy post-processing would be to expose 3d-depth and normal to canvas items.
